runtime/src/iree-amd-aie/driver/hsa/hsa_driver.cc

```diff
@@ -0,0 +1,355 @@
+      host_allocator, &driver->hsa_symbols);
+
+  IREE_RETURN_AND_END_ZONE_IF_ERROR(z0, iree_hal_hsa_init(driver));
+
+  size_t agent_index = 0;
+  iree_hal_hsa_callback_package_t symbols_and_agents = {
+      .driver = driver, .index = &agent_index, .return_value = driver->agents};
+
+  IREE_HSA_RETURN_AND_END_ZONE_IF_ERROR(
+      z0, &driver->hsa_symbols,
+      hsa_iterate_agents(&iterate_populate_aie_agents_callback,
+                         &symbols_and_agents),
+      "hsa_iterate_agents");
+
+  if (!driver->num_aie_agents) {
+    status =
+        iree_make_status(IREE_STATUS_NOT_FOUND, "no AIE agents discovered");
+  }
+
+  if (iree_status_is_ok(status)) {
+    *out_driver = reinterpret_cast<iree_hal_driver_t*>(driver);
+  } else {
+    iree_hal_driver_release(reinterpret_cast<iree_hal_driver_t*>(driver));
+  }
+
+  IREE_TRACE_ZONE_END(z0);
+  return status;
+}
+
+static void iree_hal_hsa_driver_destroy(iree_hal_driver_t* base_driver) {
+  IREE_ASSERT_ARGUMENT(base_driver);
+
+  iree_hal_hsa_driver_t* driver = iree_hal_hsa_driver_cast(base_driver);
+  iree_allocator_t host_allocator = driver->host_allocator;
+  IREE_TRACE_ZONE_BEGIN(z0);
+
+//  iree_hal_hsa_dynamic_symbols_deinitialize(&driver->hsa_symbols);
+  iree_allocator_free(host_allocator, driver);
+
+  IREE_TRACE_ZONE_END(z0);
+}
+
+static iree_hal_device_id_t iree_hsa_device_to_device_id(
+    iree_hal_hsa_driver_t* driver, hsa::hsa_agent_t agent) {
+  iree_hal_device_id_t device_id = 0;
+  while (device_id != IREE_HAL_HSA_MAX_DEVICES &&
+         driver->agents[device_id++].handle != agent.handle)
+    ;
+
+  return device_id;
+}
+
+static hsa::hsa_agent_t iree_device_id_to_hsadevice(
+    iree_hal_hsa_driver_t* driver, iree_hal_device_id_t device_id) {
+  return driver->agents[device_id];
+}
+
+static iree_status_t get_hsa_agent_uuid(iree_hal_hsa_dynamic_symbols_t* syms,
+                                        hsa::hsa_agent_t agent,
+                                        char* out_device_uuid) {
+  auto uuid_info =
+      static_cast<hsa::hsa_agent_info_t>(hsa::HSA_AMD_AGENT_INFO_UUID);
+  IREE_HSA_RETURN_IF_ERROR(
+      syms, hsa_agent_get_info(agent, uuid_info, out_device_uuid),
+      "hsa_agent_get_info");
+
+  return iree_ok_status();
+}
+
+static iree_status_t iree_hal_hsa_populate_device_info(
+    iree_hal_hsa_driver_t* driver, hsa::hsa_agent_t agent,
+    iree_hal_hsa_dynamic_symbols_t* syms, uint8_t* buffer_ptr,
+    uint8_t** out_buffer_ptr, iree_hal_device_info_t* out_device_info) {
+  *out_buffer_ptr = buffer_ptr;
+
+  char device_name[IREE_HAL_HSA_MAX_DEVICE_NAME_LENGTH];
+
+  IREE_HSA_RETURN_IF_ERROR(
+      syms, hsa_agent_get_info(agent, hsa::HSA_AGENT_INFO_NAME, device_name),
+      "hsa_agent_get_info");
+  memset(out_device_info, 0, sizeof(*out_device_info));
+
+  out_device_info->device_id = iree_hsa_device_to_device_id(driver, agent);
+
+  // Maximum UUID is 21
+  char device_uuid[21] = {0};
```

Comment:
Just wondering, where does this come from?

And maybe create a global constant for it.